tests/unit/cmor/test_cmor_check.py

```diff
@@ -918,6 +921,78 @@ def test_frequency_not_supported(self):
         """Fail at metadata if frequency is not supported."""
         self._check_fails_in_metadata(frequency='wrong_freq')
 
+    def test_hourly_time_bounds(self):
+        """Test time bounds are properly guessed for hourly freq"""
+        cube = self.get_cube(self.var_info, frequency='hr')
+        original_bounds = cube.coord('time').bounds
+        cube.coord('time').bounds = None
+        self.cube = cube
+        self._check_cube(automatic_fixes=True, frequency='hr')
+        guessed_bounds = self.cube.coord('time').bounds
+        assert original_bounds.all() == guessed_bounds.all()
+
+    def test_3hourly_time_bounds(self):
+        """Test time bounds are properly guessed for 3hourly freq"""
+        cube = self.get_cube(self.var_info, frequency='3hr')
+        original_bounds = self.cube.coord('time').bounds
+        cube.coord('time').bounds = None
+        self.cube = cube
+        self._check_cube(automatic_fixes=True, frequency='3hr')
+        guessed_bounds = self.cube.coord('time').bounds
+        assert original_bounds.all() == guessed_bounds.all()
+
+    def test_6hourly_time_bounds(self):
+        """Test time bounds are properly guessed for 6hourly freq"""
+        cube = self.get_cube(self.var_info, frequency='6hr')
+        original_bounds = self.cube.coord('time').bounds
+        cube.coord('time').bounds = None
+        self.cube = cube
+        self._check_cube(automatic_fixes=True, frequency='6hr')
+        guessed_bounds = self.cube.coord('time').bounds
+        assert original_bounds.all() == guessed_bounds.all()
+
+    def test_daily_time_bounds(self):
+        """Test time bounds are properly guessed for daily freq"""
+        original_bounds = self.cube.coord('time').bounds
+        self.cube.coord('time').bounds = None
+        self._check_cube(automatic_fixes=True)
+        guessed_bounds = self.cube.coord('time').bounds
+        assert original_bounds.all() == guessed_bounds.all()
+
+    def test_monthly_time_bounds(self):
+        """Test time bounds are guessed for monthly freq"""
+        cube = self.get_cube(self.var_info, frequency='mon')
+        cube.coord('time').bounds = None
+        self.cube = cube
+        self._check_cube(automatic_fixes=True, frequency='mon')
+        assert self.cube.coord('time').bounds is not None
```

Comment:
I noticed you check the actual values of the bounds for (sub)daily data, but from here on you just check for their presence.

If I understand correctly, this is because the test cubes for monthly and yearly data have their time points at the beginning of the interval, instead of at the midpoint, and hence the [original bounds](https://github.com/ESMValGroup/ESMValCore/blob/b632f9ac7172a91b52fe91264f37a2ce3e13ac3e/tests/unit/cmor/test_cmor_check.py#L1065) do not match your criteria. This makes me wonder: should the cmor check also verify the actual time points? And is it necessary to check the values of the bounds here, or is it good enough to see that bounds are present? 

I think a good compromise might be to move some of the tests to the unit tests for the _time module. There we could check whether the guessed bounds are actually what we expect. We could also check if it works for different calendars, (this is currently not covered, I think). Then, in the test for the cmor check, it would suffice to just check whether the bounds attribute is not None.

What do you think?

Reply:
> If I understand correctly, this is because the test cubes for monthly and yearly data have their time points at the beginning of the interval, instead of at the midpoint, and hence the original bounds do not match your criteria. This makes me wonder: should the cmor check also verify the actual time points? And is it necessary to check the values of the bounds here, or is it good enough to see that bounds are present?

Again, not sure if there is such a standard on the actual point values. The motivation behind this pull request is the fact that there is observational data in our archive that is kind of well cmorised but it's just missing the time bounds, which later makes some preprocessors crash. So adding bounds on the fly is easier than start writing fixes for each one of the datasets (which we don't know how many of those there are). But yes, if the original midpoint is not set already at the middle of the time interval (which is what happens in the test cube), the bounds will look weird for monthly, annual and decadal data. Not sure what should be done in that case though.

>We could also check if it works for different calendars, (this is currently not covered, I think).

The bounds are added taking into account the time coordinate units, in which the calendar is included, so I would say that it should be covered already. But sure, better to add tests.

Reply:
Yes I agree that it should work, but it would be good to have a test to confirm this. That ensures that if someone else later changes the code such that it will break for different calendars, the tests will point this out.